securedrop/journalist_app/api2/types.py

```diff
@@ -0,0 +1,109 @@
+from dataclasses import dataclass, field
+from enum import IntEnum, StrEnum, auto
+from typing import (
+    Any,
+    List,
+    NewType,
+    Optional,
+    Set,
+    Tuple,
+    Union,
+)
+
+Record = NewType("Record", dict[str, Any])
+Version = NewType("Version", str)
+
+
+# TODO: generic UUID[T] in Python 3.12
+SourceUUID = NewType("SourceUUID", str)
+ItemUUID = NewType("ItemUUID", str)
+JournalistUUID = NewType("JournalistUUID", str)
+
+
+EventID = NewType("EventID", str)  # int, but opaque on the wire
+
+
+class EventType(StrEnum):
+    REPLY_SENT = auto()
+    ITEM_DELETED = auto()
+
+
+class EventStatusCode(IntEnum):
+    OK = 200
+    AlreadyReported = 208
+    BadRequest = 400
+    NotFound = 404
+    NotImplemented = 501
+
+
+EventStatus = Tuple[EventStatusCode, Optional[str]]
+
+
+@dataclass
+class Index:
+    # Source metadata, optionally filtered by `source_prefix`:
+    sources: dict[SourceUUID, Version] = field(default_factory=dict)
+    items: dict[ItemUUID, Version] = field(default_factory=dict)
+
+    # Non-source metadata (always returned):
+    journalists: dict[JournalistUUID, Version] = field(default_factory=dict)
+
+
+@dataclass
+class SourceTarget:
+    source_uuid: SourceUUID
+    version: Version
+
+
+@dataclass
+class ItemTarget:
+    item_uuid: ItemUUID
+    version: Version
+
+
+@dataclass
+class Event:
+    id: EventID
+    target: Union[SourceTarget, ItemTarget]
+    type: EventType
+    data: dict[str, Any] = field(default_factory=dict)
+
+
+@dataclass
+class EventResult:
+    event_id: EventID
+    status: EventStatus
+
+    # Changed (return {<uuid>: None} to indicate deletion):
```

Comment:
nit: the content in the parenthetical is important and not an aside :)

```suggestion
    # Changed sources/items, return {<uuid>: None} to indicate deletion:
```

Reply:
This is also explicit here:

https://github.com/freedomofpress/securedrop/blob/6bebcfb4f2f0014d5ec836492d61081f90add4c8/securedrop/journalist_app/api2/types.py#L96-L98